### ISSUE TITLE
possible broken installation

### ISSUE THREAD

**Issue by kingrodriguez** [calebstewart/pwncat#217]
## Bug Description
the command `report` is shown in the modules (`search *`).
it is stated as unknown command / throws an error.
i am a little unsure if directly running a module is the way or `run $module`

## pwncat version
Provide the output of `pwncat --version` or a commit hash if working from
a development branch.

```
$ pwncat --version
0.4.3
```

## Target System (aka "victim")
self / none

## Steps to Reproduce
1. `python -m pwncat`
2. `pwncat$ report`
3. `error: report: unknown command `
4. `pwncat$ run report`
5. `ValueError: The 'pwncat' package was not installed in a way that PackageLoader 
understands.`

result (with stack trace)

```
pwncat$ run report

╭──────────────────────── Traceback (most recent call last) ────────────────────────╮
│                                                                                   │
│ /usr/lib/python3.9/site-packages/pwncat/commands/__init__.py:586 in run           │
│                                                                                   │
│    583 │   │   │   │   if line == "":                                             │
│    584 │   │   │   │   │   continue                                               │
│    585 │   │   │   │                                                              │
│ ❱  586 │   │   │   │   self.dispatch_line(line)                                   │
│    587 │   │   │   # We used to catch only KeyboardException, but this prevents a │
│    588 │   │   │   # badly written command from completely killing our remote     │
│    589 │   │   │   # connection.                                                  │
│ /usr/lib/python3.9/site-packages/pwncat/commands/__init__.py:667 in dispatch_line │
│                                                                                   │
│    664 │   │   │   │   args = line                                                │
│    665 │   │   │                                                                  │
│    666 │   │   │   # Run the command                                              │
│ ❱  667 │   │   │   command.run(self.manager, args)                                │
│    668 │   │   │                                                                  │
│    669 │   │   │   if prog_name:                                                  │
│    670 │   │   │   │   command.parser.prog = prog_name                            │
│                                                                                   │
│ /usr/lib/python3.9/site-packages/pwncat/commands/run.py:67 in run                 │
│                                                                                   │
│    64 │   │   config_values.update(values)                                        │
│    65 │   │                                                                       │
│    66 │   │   try:                                                                │
│ ❱  67 │   │   │   result = manager.target.run(module_name, **config_values)       │
│    68 │   │   │                                                                   │
│    69 │   │   │   if args.module is not None:                                     │
│    70 │   │   │   │   manager.config.back()                                       │
│                                                                                   │
│ /usr/lib/python3.9/site-packages/pwncat/manager.py:200 in run                     │
│                                                                                   │
│   197 │   │   if module.PLATFORM is not None and type(self.platform) not in modul │
│   198 │   │   │   raise pwncat.modules.IncorrectPlatformError(module_name)        │
│   199 │   │                                                                       │
│ ❱ 200 │   │   return module.run(self, **kwargs)                                   │
│   201 │                                                                           │
│   202 │   def find_module(self, pattern: str, base=None, exact: bool = False):    │
│   203 │   │   """Locate a module by a glob pattern. This is an generator          │
│                                                                                   │
│ /usr/lib/python3.9/site-packages/pwncat/modules/__init__.py:232 in decorator      │
│                                                                                   │
│   229 │   │   try:                                                                │
│   230 │   │   │                                                                   │
│   231 │   │   │   # Return the result                                             │
│ ❱ 232 │   │   │   result_object = real_run(self, session, **kwargs)               │
│   233 │   │   │                                                                   │
│   234 │   │   │   if inspect.isgenerator(result_object):                          │
│   235 │   │   │   │   if session.showing_progress:                                │
│                                                                                   │
│ /usr/lib/python3.9/site-packages/pwncat/modules/agnostic/report.py:87 in run      │
│                                                                                   │
│    84 │   │   │   )                                                               │
│    85 │   │   else:                                                               │
│    86 │   │   │   env = jinja2.Environment(                                       │
│ ❱  87 │   │   │   │   loader=jinja2.PackageLoader("pwncat", "data/reports"),      │
│    88 │   │   │   │   # autoescape=jinja2.select_autoescape(["md", "html"]),      │
│    89 │   │   │   │   trim_blocks=True,                                           │
│    90 │   │   │   │   lstrip_blocks=True,                                         │
│                                                                                   │
│ /usr/lib/python3.9/site-packages/jinja2/loaders.py:311 in __init__                │
│                                                                                   │
│   308 │   │   │   │   │   break                                                   │
│   309 │   │                                                                       │
│   310 │   │   if template_root is None:                                           │
│ ❱ 311 │   │   │   raise ValueError(                                               │
│   312 │   │   │   │   f"The {package_name!r} package was not installed in a"      │
│   313 │   │   │   │   " way that PackageLoader understands."                      │
│   314 │   │   │   )                                                               │
╰───────────────────────────────────────────────────────────────────────────────────╯
ValueError: The 'pwncat' package was not installed in a way that PackageLoader 
understands.
```

## Expected Behavior
execute the command / not throwing an error

**Comment by DanaEpp** [comment]
I have this same issue.

### Affected System
- Linux kali 5.14.0-kali2-amd64 #1 SMP Debian 5.14.9-2kali1 (2021-10-04) x86_64 GNU/Linux
- Pwncat: 0.4.3
- Python: 3.9.7

### Traceback
```
(local) pwncat$ run report                                                                                                          
╭─────────────────────────────────────────────── Traceback (most recent call last) ────────────────────────────────────────────────╮
│ /usr/local/lib/python3.9/dist-packages/pwncat-0.4.3-py3.9.egg/pwncat/commands/__init__.py:586 in run                             │
│                                                                                                                                  │
│    583 │   │   │   │   if line == "":                                                                                            │
│    584 │   │   │   │   │   continue                                                                                              │
│    585 │   │   │   │                                                                                                             │
│ ❱  586 │   │   │   │   self.dispatch_line(line)                                                                                  │
│    587 │   │   │   # We used to catch only KeyboardException, but this prevents a                                                │
│    588 │   │   │   # badly written command from completely killing our remote                                                    │
│    589 │   │   │   # connection.                                                                                                 │
│                                                                                                                                  │
│ /usr/local/lib/python3.9/dist-packages/pwncat-0.4.3-py3.9.egg/pwncat/commands/__init__.py:667 in dispatch_line                   │
│                                                                                                                                  │
│    664 │   │   │   │   args = line                                                                                               │
│    665 │   │   │                                                                                                                 │
│    666 │   │   │   # Run the command                                                                                             │
│ ❱  667 │   │   │   command.run(self.manager, args)                                                                               │
│    668 │   │   │                                                                                                                 │
│    669 │   │   │   if prog_name:                                                                                                 │
│    670 │   │   │   │   command.parser.prog = prog_name                                                                           │
│                                                                                                                                  │
│ /usr/local/lib/python3.9/dist-packages/pwncat-0.4.3-py3.9.egg/pwncat/commands/run.py:67 in run                                   │
│                                                                                                                                  │
│    64 │   │   config_values.update(values)                                                                                       │
│    65 │   │                                                                                                                      │
│    66 │   │   try:                                                                                                               │
│ ❱  67 │   │   │   result = manager.target.run(module_name, **config_values)                                                      │
│    68 │   │   │                                                                                                                  │
│    69 │   │   │   if args.module is not None:                                                                                    │
│    70 │   │   │   │   manager.config.back()                                                                                      │
│                                                                                                                                  │
│ /usr/local/lib/python3.9/dist-packages/pwncat-0.4.3-py3.9.egg/pwncat/manager.py:200 in run                                       │
│                                                                                                                                  │
│   197 │   │   if module.PLATFORM is not None and type(self.platform) not in module.PLATFORM:                                     │
│   198 │   │   │   raise pwncat.modules.IncorrectPlatformError(module_name)                                                       │
│   199 │   │                                                                                                                      │
│ ❱ 200 │   │   return module.run(self, **kwargs)                                                                                  │
│   201 │                                                                                                                          │
│   202 │   def find_module(self, pattern: str, base=None, exact: bool = False):                                                   │
│   203 │   │   """Locate a module by a glob pattern. This is an generator                                                         │
│                                                                                                                                  │
│ /usr/local/lib/python3.9/dist-packages/pwncat-0.4.3-py3.9.egg/pwncat/modules/__init__.py:232 in decorator                        │
│                                                                                                                                  │
│   229 │   │   try:                                                                                                               │
│   230 │   │   │                                                                                                                  │
│   231 │   │   │   # Return the result                                                                                            │
│ ❱ 232 │   │   │   result_object = real_run(self, session, **kwargs)                                                              │
│   233 │   │   │                                                                                                                  │
│   234 │   │   │   if inspect.isgenerator(result_object):                                                                         │
│   235 │   │   │   │   if session.showing_progress:                                                                               │
│                                                                                                                                  │
│ /usr/local/lib/python3.9/dist-packages/pwncat-0.4.3-py3.9.egg/pwncat/modules/agnostic/report.py:87 in run                        │
│                                                                                                                                  │
│    84 │   │   │   )                                                                                                              │
│    85 │   │   else:                                                                                                              │
│    86 │   │   │   env = jinja2.Environment(                                                                                      │
│ ❱  87 │   │   │   │   loader=jinja2.PackageLoader("pwncat", "data/reports"),                                                     │
│    88 │   │   │   │   # autoescape=jinja2.select_autoescape(["md", "html"]),                                                     │
│    89 │   │   │   │   trim_blocks=True,                                                                                          │
│    90 │   │   │   │   lstrip_blocks=True,                                                                                        │
│                                                                                                                                  │
│ /usr/lib/python3/dist-packages/jinja2/loaders.py:309 in __init__                                                                 │
│                                                                                                                                  │
│   306 │   │   │   │   │   break                                                                                                  │
│   307 │   │                                                                                                                      │
│   308 │   │   if template_root is None:                                                                                          │
│ ❱ 309 │   │   │   raise ValueError(                                                                                              │
│   310 │   │   │   │   f"The {package_name!r} package was not installed in a"                                                     │
│   311 │   │   │   │   " way that PackageLoader understands."                                                                     │
│   312 │   │   │   )                                                                                                              │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
ValueError: The 'pwncat' package was not installed in a way that PackageLoader understands.
```

**Comment by calebstewart** [comment]
Hm, this appears to be a problem with the Jinja2 `PackageLoader` (or at least an incompatibility between `PackageLoader` and installing via `pip install git+...`). I am not sure why Jinja2 has randomly decided to fail. I can reproduce this locally with a new virtualenv, but not in my development virtualenv with the same version of `Jinja2`.

I'll look into this, but I'm not sure what the problem is yet.

**Comment by calebstewart** [comment]
I've tested installing with `pip install git+...` like in the docs as well as `git clone ... && pip install ./pwncat` and `git clone ... && cd pwncat && python setup.py install` and they all reproduce this, sadly.

**Comment by DanaEpp** [comment]
Is it possible the Jinga2 versions differ in the new virtualenv and your dev virtualenv? 

Mine is showing:
```bash
$ pip list | grep Jinja2   
Jinja2                       3.0.1
```

**Comment by DanaEpp** [comment]
> I've tested installing with `pip install git+...` like in the docs as well as `git clone ... && pip install ./pwncat` and `git clone ... && cd pwncat && python setup.py install` and they all reproduce this, sadly.

Ya, I used the git clone -> python setup.py install method for my setup. And I didn't install it into a virtualenv.

**Comment by calebstewart** [comment]
I've been trying to clean up this repo today since I've been distracted/busy for a bit and have let things go. I released `v0.5.0` a bit ago, which switched to using Python Poetry for dependency management, and also published the project to PyPI under the name `pwncat-cs`. Could you possibly try installing this new version (either by cloning and running `pip install ./pwncat` or by just doing `pip install pwncat-cs` to install from PyPI)?

**NOTE** - the name of the entrypoint has changed to `pwncat-cs` to deconflict with Cytopia's pwncat command. :+1: 

I can't reproduce this now with the updates/changes there, so I think it may have been an issue with the packaging done by `setup.py`. Using `setup.py` isn't recommended anymore anyway, so I'm hoping that moving to Poetry just had the side-effect of fixing this. lol

**Comment by DanaEpp** [comment]
Hey Caleb,

So I did an install using pip:
```bash
$ pip3 install pwncat-cs
$ pwncat-cs -v
0.4.3
```

I then ran an active session to test reports (and new features you added for lcd/lpwd):
```bash
$ pwncat-cs -lp 4242

(local) pwncat$ lpwd
/home/silverstr
(local) pwncat$ lcd /tmp
(local) pwncat$ lpwd
/tmp
(local) pwncat$ lcd -
usage: lcd [-h] path
lcd: error: -: not a directory
(local) pwncat$ lcd ~
(local) pwncat$ lpwd
/home/silverstr
(local) pwncat$ run enumerate
[REDACTED - BUT SUCCESSFUL]
(local) pwncat$ run report
[REDACTED - BUT SUCCESSFUL]
(local) pwncat$ exit
```

So it appears the package is fine to run reports in pwncat-cs. And the new features are working fine as well, which is sweet. Thanks for adding them in for me.

My only question now is the versioning. You mentioned it was 0.5.0. However, the latest install for me shows as 0.4.3 with the new features in. The change log shows it was added in 0.4.4, and if 0.5.0 is the latest, I wonder why it installed 0.4.3 for me. 

Any ideas?

**Comment by calebstewart** [comment]
The version showing incorrectly is in fact a bug that I'll fix right now. It's reporting the version of the `pwncat` package instead of `pwncat-cs`, which is my bad. If you look at `pip freeze | grep "pwncat"`, you should see `v0.5.0`. The offending lines are in `__main__.py` at line 103. It should be doing `importlib.metadata.version("pwncat-cs")`.

```py
    # Print the version number and exit.
    if args.version:
        print(importlib.metadata.version("pwncat"))
        return
```

There were some typos in the README as well, so I'll just push these small fixes as `0.5.1` tomorrow after I spend some more time trying to find any other small typos like this.

**Comment by calebstewart** [comment]
I'm going to close this issue as it seems the original problem is fixed, and the version issue is fixed in the latest release. :+1:
